First draft of view_user_birthdate complete with test modules and tex…

CoitThomas opened the pull request:
First draft of view_user_birthdate complete with test modules and text files.

CoitThomas committed:
First draft of view_user_birthdate complete with test modules and text files.

austinwiltshire commented on view_user_birthdate.py:
```diff
@@ -0,0 +1,242 @@
+        if self.username not in User.bday_book:
+            User.bday_book[username] = self.birthdate
+
+class Date(object):
```
read https://pymotw.com/2/abc/

and add 'format' as an abstract method here.

austinwiltshire commented on view_user_birthdate.py:
```diff
@@ -0,0 +1,242 @@
+        string in the American format:
+        mm/dd/yyyy
+        """
+        return '%s/%s/%s' % (self.month, self.day, self.year)
```
don't refer to a parent classes variables - try to access only via functions. in this case, you'd want to write getters for month, day and year.

read: http://stackoverflow.com/questions/1641219/does-python-have-private-variables-in-classes

and hide all of your instance variables using the _leading_underscore approach. Pylint will enforce privacy then.

austinwiltshire commented on view_user_birthdate.py:
```diff
@@ -0,0 +1,242 @@
+    assert 0 < int(size) <= len(data_lines), "A valid size must be provided."
+    assert len(data_lines) % size == 0, "Data is incomplete."
+
+    return [data_lines[pos:pos + size] for pos in range(0, len(data_lines), size)]
```
very nice.

austinwiltshire commented on view_user_birthdate.py:
```diff
@@ -0,0 +1,242 @@
+    """
+    # Parse username.
+    first_line = re.search(r"username:(?P<username>[\w.-]+)", data_chunk[0])
+    if first_line:
```
this repeats a common pattern - can you refactor this into a common function, something like the signature

def parse_chunk(regex, regex_group, chunk)

ex:

parse_chunk(r"username: ...", 'username', 0)

austinwiltshire commented on view_user_birthdate.py:
```diff
@@ -0,0 +1,242 @@
+            birthdate = format_date(month, day, year, nationality)
+            # Instantiate a User object thus adding the User to the bday_book.
+            User(username, nationality, birthdate)
+    return User.bday_book
```
you're breaking encapsulation here, and also referring to a class variable. In python, anything declared at class level is at the class level - you create instance/object variables by setting them up in the **init** function, self.bday_book, etc...

austinwiltshire commented on view_user_birthdate.py:
```diff
@@ -0,0 +1,242 @@
+            month, day, year = build_date(birthdate)
+            birthdate = format_date(month, day, year, nationality)
+            # Instantiate a User object thus adding the User to the bday_book.
+            User(username, nationality, birthdate)
```
constructors shouldn't have side effects - here you're creating a User object then not doing anything with it, entirely relying on the side effect. it'd be better to do something like

bday_book[username] = User(username, nationality, birthdate)

The bday book dictionary can be built outside of user, doesn't need to live inside user.

austinwiltshire commented on view_user_birthdate.py:
```diff
@@ -0,0 +1,242 @@
+    year = make_int(date.group('year'))
+
+    if is_valid_date(month, day, year):
+        return str(month), str(day), str(year)
```
what does this function do if is_valid_date fails?

austinwiltshire commented on view_user_birthdate.py:
```diff
@@ -0,0 +1,242 @@
+    assert valid_day(make_int(day), make_int(month)), "Invalid day."
+    assert valid_year(make_int(year)), "Invalid year."
+    assert re.search("[a-zA-Z]+ ?[a-zA-Z]*", nationality), "Invalid nationality."
+    if nationality.lower() in AmericanDate.nationalities:
```
this is not cohesive. Why would the idea of nationalities live inside the idea of an AmericanDate?

austinwiltshire commented on view_user_birthdate.py:
```diff
@@ -0,0 +1,242 @@
+    assert re.search("[a-zA-Z]+ ?[a-zA-Z]*", nationality), "Invalid nationality."
+    if nationality.lower() in AmericanDate.nationalities:
+        return AmericanDate(month, day, year).format()
+    if nationality.lower() in EuropeanDate.nationalities:
```
good job on thinking of the lower. Also i'm assuming you expanded the nationalities supported just for flair and are aware it violates KISS in the real world

austinwiltshire commented on view_user_birthdate.py:
```diff
@@ -0,0 +1,242 @@
+    if nationality.lower() in EuropeanDate.nationalities:
+        return EuropeanDate(month, day, year).format()
+    if nationality.lower() in ChineseDate.nationalities:
+        return ChineseDate(month, day, year).format()
```
consider holding on to the actual date objects longer, and only format them when you need to. You may need to add other functionality to date later, and that'd require a refactor since right now you're passing around strings everywhere instead of the object they derived from.

austinwiltshire commented on view_user_birthdate.py:
```diff
@@ -0,0 +1,242 @@
+    """Create an object of type ChineseDate wich inherits from class
+    Date.
+    """
+    nationalities = ['chinese', 'north korean', 'south korean',
```
as mentioned below, the idea of nationalities don't belong in the date classes.

austinwiltshire replied:
this is also a class variable which it doesn't make sense as right now

austinwiltshire commented on view_user_birthdate.py:
```diff
@@ -0,0 +1,242 @@
+        return '%s/%s/%s' % (self.month, self.day, self.year)
+
+class EuropeanDate(Date):
+    """Create an object of type EuropeanDate which inherits from class
```
docs are too wordy here. this isn't telling me anything the code is not - tell me how eurpean dates differ from other dates?

austinwiltshire commented on view_user_birthdate.py:
```diff
@@ -0,0 +1,242 @@
+        assert valid_month(make_int(month)), "Invalid month."
+        assert valid_day(make_int(day), make_int(month)), "Invalid day."
+        assert valid_year(make_int(year)), "Invalid year."
+        self.month = month
```
use the _month and _day syntax to encapsulate these variables, then add month() and day() accessors.

austinwiltshire replied:
_alternatively_, in a way that avoids accessors cause those are gross too, consider the template method pattern:

https://en.wikipedia.org/wiki/Template_method_pattern

In this case, your base class, date, would support a _non abstract method_ called format(self), defined similar to the following:

def format(self):
    """ Calls on the child classes's formatting protocol to get the month/day/year right """
    return self._format(self.month, self.day, self.year)

In addition, Date class will have  a _format(self, month, day, year) _abstract method_. 

@abstractmethod
def _format(self, month, day, year):
    pass

each child class will _define _format_ method in a way that makes sense for them

class AmericanDate(Date):
def _format(self, month, day, year):
   return the american date

austinwiltshire commented on view_user_birthdate.py:
```diff
@@ -0,0 +1,242 @@
+    variables as attributes. Keep track of all instantiated User
+    objects with the dictionary 'bday_book'.
+    """
+    bday_book = {}
```
this class variable can't be reused- if i create a User anywhere in my system, the _global bday_book variable_ gets updated, no matter where I am. this means parts of my program that have nothing to do with each other are _coupled_ needlessly. move this guy out

austinwiltshire commented on view_user_birthdate.py:
```diff
@@ -0,0 +1,242 @@
+        """
+        assert re.search(r"[\w.-]+", username), "Invalid username."
+        assert re.search("[a-zA-Z]+ ?[a-zA-Z]*", nationality), "Invalid nationality."
+        assert re.search("[0-9]{1,4}/[0-9]{1,2}/[0-9]{1,4}", birthdate)
```
i'd drop this requirement on birthdate and pull in a Date class.

austinwiltshire commented on view_user_birthdate.py:
```diff
@@ -0,0 +1,242 @@
+    """
+    bday_book = {}
+
+    def __init__(self, username, nationality, birthdate):
```
research the **str** and **repr** magic methods in python, and use those to print your user rather than passing strings around.  this will allow you to pass around your User objects, but when you do print User('a','b','c') it will give back a formatted print string however you like

Do the same with date, _put the call to format in the **repr** method_

austinwiltshire commented on test_build_bday_book.py:
```diff
@@ -0,0 +1,13 @@
+def test_build_bday_book():
+    """Assert correct contents of an assmbled bday_book."""
+    file_name = 'testfile4build_bday_book.txt'
+    assert view_user_birthdate.build_bday_book(file_name) == {'austinwiltshire':'4/20/1983',
```
I want your dictionary to be from usernames to users, and your users to print smartly enough that they print the dates right, don't precalculate the dates.

Your design would not survive me reassigning someone's nationality, for instance, since i forgot what they originally were. If you keep the users around and format at the last second, you keep information around that makes future changes and use easier.

austinwiltshire commented:
This is a lot of code, lots of tests, great assertions as always. And I think its a great first run at OO.

CoitThomas committed:
Refactor parse().

CoitThomas committed:
Refactor build_date().

CoitThomas committed:
Issue

CoitThomas committed:
Refactor User class and Date class. Implement date_format_book dictionary. Remove AmericanDate class.

CoitThomas committed:
Make class attributes private.

CoitThomas committed:
Refactor unit tests.

austinwiltshire commented:
I am commenting on a per commit basis so I don't accidentally recomment. Then i'll comment on the whole thing again for any lingering issues.

I like your comments so far! I'll bet 6 months ago if I told you to refactor the unit tests, you'd be very confused.

austinwiltshire commented on view_user_birthdate.py:
```diff
@@ -0,0 +1,251 @@
+        assert re.search(REGEX_DATE_RULE, string), assert_response
+        self._month, self._day, self._year = parse_date(string)
+
+    def get_date_attributes(self):
```
for the template method pattern i mentioned elsewhere to prevent this getter, you'd write something like

```
def display_date(self):
    return self._format_date(self._month, self._day, self._year)

@abstractmethod
def _format_date(self, month, day, year):
    raise NotImplementedError()
```

you'd write the above in Date.

In American, European and ChineseDate, you'd write

```
def _format_date(self, month, day, year):
   ...do the formatting logic here...
```

In this way, the parent class never has its private variables 'gotten'. It instead 'tells' its children by calling them and _providing_ its private variables. This has at least two benefits:
1. Children don't know how the parent is implemented (reduced coupling)
2. Children are easier to test since a test can just pass dummy values in

This pattern is also close to the 'strategy pattern' you might be interested in reading up on.

austinwiltshire replied:
In the above, you can see I made a private method - this is fine for _parents_ to call their _children's_ private methods. The reverse is not true, and in either case of private variables, that also violates encapsulation.

The parent, through the 'abstractmethod' decorator, is demanding that anyone who implements it _must_ provide that function. It's a kind of contract, though not the same as assertions. Since it's demanding all children implement that function, the parent is safe to call that function and does not violate encapsulation by doing so.

CoitThomas committed:
Create a private template method pattern _format_date().
Add AmericanDate class.
Remove getter methods from User class.
Adjust parameter requirements for Date objects.
Create show_self() method for Date class.
Simplify parse().
Improve readability of build_bday_book().
Create User unit test.

CoitThomas committed:
Remove getter method 'show_self()' from Date class.
Remove format_birthdate() method from User class.
Assign nationalized birthdate when User object is instantiated.
Remove User unit test.
Remove Date unit test.

CoitThomas committed:
Correct exception typo in _format_date() method of Date class.
